alternator/consumed_capacity.cc

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "consumed_capacity.hh"
+#include "error.hh"
+#include <seastar/core/metrics.hh>
+
+namespace alternator {
+
+/*
+ * \brief DynamoDB counts read capacity in half-integers - a short
+ * eventually-consistent read is counted as 0.5 unit.
+ * Because we want our counter to be an integer, it counts half units.
+ * Both read and write counters count in these half-units, and should be
+ * multiply by 0.5 (INTERNAL_UNIT_MULTIPLIER) to get the DynamoDB-compatible RCU or WCU numbers.
+ */
+static constexpr double INTERNAL_UNIT_MULTIPLIER = 0.5;
+
+static constexpr uint64_t KB = 1024ULL;
+static constexpr uint64_t RCU_BLOCK_SIZE_LENGTH = 4*KB;
+static constexpr uint64_t WCU_BLOCK_SIZE_LENGTH = 1*KB;
+
+static bool should_add_capacity(const rjson::value& request) {
+    const rjson::value* return_consumed = rjson::find(request, "ReturnConsumedCapacity");
+    if (!return_consumed) {
+        return false;
+    }
+    if (!return_consumed->IsString()) {
+        throw api_error::validation("Non-string ReturnConsumedCapacity field in request");
+    }
+    std::string consumed = return_consumed->GetString();
+    if (consumed == "INDEXES") {
+        throw api_error::validation("INDEXES consumed capacity is not supported");
+    }
+    return consumed == "TOTAL";
+}
+
+void consumed_capacity_counter::add_consumed_capacity_to_response_if_needed(rjson::value& response) const noexcept {
+    if (_should_add_to_reponse) {
+        auto consumption = rjson::empty_object();
+        rjson::add(consumption, "CapacityUnits", get_consumed_capacity_units());
+        rjson::add(response, "ConsumedCapacity", std::move(consumption));
+    }
+}
+
+void consumed_capacity_counter::update_metric(replica::consumption_unit_counter& metric) const noexcept {
+    metric.add_units(get_internal_units());
+}
+
+static uint64_t calculate_internal_units(uint64_t unit_block_size, uint64_t total_bytes, bool is_quorum) {
+    uint64_t internal_units = (total_bytes + unit_block_size -1) / unit_block_size; //divide by unit_block_size and round up
+
+    if (is_quorum) {
+        internal_units *= 2;
```

Comment:
We should measure on the replica, this way we account for all the replicas, even if RF=7.

Reply:
What you describe is the internal usage. This is not the same as the RCU and WCU

Reply:
The replica is running low-level query functions for us, and doesn't know that it is performing a piece of work for Alternator, so can't log this.
Also please note that in Alternator (the DynamoDB API) you always have RF=3 and read CL is either ONE or QUORUM (in DynamoDB API this corresponds to eventually-consistent and strongly-consistent reads, respectively). Weird RF and CL is not something that can happen. Even if we ever do allow to configure them (https://github.com/scylladb/scylladb/issues/7428), I don't know what we would expect the ReturnConsumedCapacity to return in this case - there is no Amazon precedent of configuring different RF or CL.

Reply:
> The replica is running low-level query functions for us, and doesn't know that it is performing a piece of work for Alternator, so can't log this. 

It's not important if it's for alternator or not. It's a table property, not a front-end property.

> Also please note that in Alternator (the DynamoDB API) you always have RF=3 and read CL is either ONE or QUORUM (in DynamoDB API this corresponds to eventually-consistent and strongly-consistent reads, respectively). Weird RF and CL is not something that can happen. Even if we ever do allow to configure them (#7428), I don't know what we would expect the ReturnConsumedCapacity to return in this case - there is no Amazon precedent of configuring different RF or CL.

It doesn't make sense to charge for 3 replicas if there are 15, you don't need precedents for this.

Reply:
RCU and WCU are alternator concepts; in the future, we may alter them from DynamoDB, but the coordinator should do the calculation. Later, when we add usage restrictions, it will be up to the coordinator to block requests when receiving them.